### ISSUE TITLE
Are Guards Needed in Absence of Sharing?

### ISSUE THREAD

**Issue by rachitnigam** [cucapra/filament#37]
For all the data ports, why do we even need a guard? In absence of sharing, the port will either receive a `0` (because of how calyx generated muxes) or always receive the same input. If data ports behave as described in https://github.com/cucapra/calyx/issues/1169, then we can just always forward the input because the Filament type system guarantees that we will not read outputs corresponding to `'x`.

In other words, because the default assignments for data ports should be `'x` (according to the Calyx proposal above), it is safe to replace the assignment: `in = guard ? data : 'x` with `in = data` because `'x` can be any value.

**Comment by rachitnigam** [comment]
It occurs to me that this optimization should be done by Calyx and not Filament itself because this reasoning applies to all data ports. Specifically, rewriting:
```
in = guard ? data : 'x
```
into
```
in = data
```
Should always be safe

**Comment by rachitnigam** [comment]
Actually, I think this line of reasoning renders the pipelined FSM generation completely useless in absence of resource sharing because we can just always forward data without checking which state is active

**Comment by rachitnigam** [comment]
Guards are required when a subcomponent requires evidence for an event. Specifically, `Register` requires evidence of the start event in the form of the `write_en` signal so it is not possible to eliminate `go` signals from sequential designs. We need something like #44 for that

**Comment by rachitnigam** [comment]
Guards are not need but FSMs still are. Phantom events represent this trade-off
